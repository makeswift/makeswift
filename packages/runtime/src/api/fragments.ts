import { DocumentNode } from 'graphql'
import { APIResourceType } from './types'
import * as FragmentDocumentNodes from './graphql/generated/fragment-document-nodes'

export const Fragments: Record<APIResourceType, DocumentNode> = FragmentDocumentNodes
