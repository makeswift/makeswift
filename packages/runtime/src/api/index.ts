export * from './types'
export * from './fragments'
