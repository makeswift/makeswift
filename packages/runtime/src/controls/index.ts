export * from './control'

export {
  ControlDefinition,
  Checkbox,
  CheckboxDefinition,
  Color,
  ColorDefinition,
  Combobox,
  ComboboxDefinition,
  Font,
  FontDefinition,
  Group,
  GroupDefinition,
  GroupControl,
  IconRadioGroupDefinition,
  type IconRadioGroupIcon,
  unstable_IconRadioGroup,
  Image,
  ImageDefinition,
  List,
  ListDefinition,
  ListControl,
  Number,
  NumberDefinition,
  Select,
  SelectDefinition,
  Shape,
  ShapeDefinition,
  ShapeControl,
  Style,
  StyleDefinition,
  type StyleProperty,
  StyleControl,
  TextInput,
  TextInputDefinition,
  TextArea,
  TextAreaDefinition,
  unstable_Typography,
  unstable_TypographyDefinition,
} from '@makeswift/controls'

export { Link, LinkDefinition } from './link'
export { RichTextV1Definition, RichTextV1Control } from './rich-text'
export { RichText, RichTextV2Definition, RichTextV2Control } from './rich-text-v2'
export { Slot, SlotDefinition, SlotControl } from './slot'
export { unstable_StyleV2, StyleV2Definition, StyleV2Control } from './style-v2/style-v2'
