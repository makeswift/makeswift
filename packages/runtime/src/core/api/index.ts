/**
 * Re-export all API-related types and utilities
 */
export * from './client';
export * from './types';