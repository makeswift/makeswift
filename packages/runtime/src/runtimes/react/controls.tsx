import { useRef, ReactNode } from 'react'

import { ControlDefinition } from '@makeswift/controls'

import { partitionRecord } from '../../utils/partition'

import * as ReactPage from '../../state/react-page'
import { isLegacyDescriptor, LegacyDescriptor } from '../../prop-controllers/descriptors'

import { useStore } from './hooks/use-store'

import { resolveLegacyDescriptorProp } from './legacy-controls'
import { useResolvedProps } from './hooks/use-resolved-props'

type PropsValueProps = {
  element: ReactPage.ElementData
  children(props: Record<string, unknown>): ReactNode
}

function useControlDefs(
  elementType: string,
): readonly [Record<string, LegacyDescriptor>, Record<string, ControlDefinition>] {
  const store = useStore()
  const all = ReactPage.getComponentPropControllerDescriptors(store.getState(), elementType) ?? {}
  return useRef(partitionRecord(all, isLegacyDescriptor)).current
}

export function ResolveProps({ element, children: renderComponent }: PropsValueProps): ReactNode {
  const [legacyDescriptors, definitions] = useControlDefs(element.type)

  const resolvedProps = useResolvedProps(definitions, element.props, element.key)

  const renderFn = Object.entries(legacyDescriptors).reduceRight(
    (renderFn, [propName, descriptor]) =>
      props =>
        resolveLegacyDescriptorProp(descriptor, propName, element.props[propName], props, renderFn),
    renderComponent,
  )

  return renderFn(resolvedProps)
}
