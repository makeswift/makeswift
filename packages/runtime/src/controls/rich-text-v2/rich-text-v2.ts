import { ReactNode } from 'react'
import { z } from 'zod'
// @ts-expect-error: there are no types for 'corporate-ipsum'
import ipsum from 'corporate-ipsum'

import {
  RichTextDefinition as BaseRichTextDefinition,
  isNotNil,
  ControlDefinition,
  SerializationSchema,
  serialize,
  StableValue,
  type Data,
  type Resolvable,
  type SendMessage,
  type SerializedRecord,
  type DeserializedRecord,
  type SchemaType,
  type SchemaTypeAny,
  type MergeTranslatableDataContext,
  type RichTextPluginControl,
  type ResourceResolver,
  type Stylesheet,
  type RichTextMode,
} from '@makeswift/controls'

import { LinkPlugin } from '../../slate/LinkPlugin'
import { InlinePlugin } from '../../slate/InlinePlugin'
import { TextAlignPlugin } from '../../slate/TextAlignPlugin'
import { BlockPlugin } from '../../slate/BlockPlugin'
import { TypographyPlugin } from '../../slate/TypographyPlugin'
import { InlineModePlugin } from '../../slate/InlineModePlugin'
import { toText } from '../../slate/utils'

import { renderRichTextV2 } from '../../runtimes/react/controls/rich-text-v2'

import { RichTextV2Plugin, Plugin } from './plugin'
import { RichTextV2Control } from './control'
import {
  getTranslatableData,
  mergeTranslatedNodes,
  type RichTextTranslationDto,
} from './translation'

type DataType = z.infer<typeof Definition.schema.data>
type DataV2Type = z.infer<typeof Definition.schema.dataV2>
type InstanceType = RichTextV2Control
type UserConfig = z.infer<typeof Definition.schema.userConfig>
type Config = UserConfig & {
  defaultValue: string
  plugins: RichTextV2Plugin[]
}

class Definition extends BaseRichTextDefinition<ReactNode, Config, InstanceType> {
  constructor({ mode, defaultValue }: UserConfig, plugins?: RichTextV2Plugin[]) {
    super({
      mode,
      defaultValue:
        defaultValue ??
        (mode === Definition.Mode.Inline ? 'Edit this text' : Definition.generateParagraph()),
      plugins:
        plugins ??
        (mode === Definition.Mode.Inline
          ? [InlineModePlugin()]
          : [BlockPlugin(), TypographyPlugin(), TextAlignPlugin(), InlinePlugin(), LinkPlugin()]),
    })
  }

  static generateParagraph(): string {
    return ipsum(3)
  }

  static deserialize(
    data: DeserializedRecord,
    deserializeCallback: (r: DeserializedRecord) => ControlDefinition,
  ): RichTextV2Definition {
    if (data.type !== Definition.type) {
      throw new Error(`RichText: expected type ${Definition.type}, got ${data.type}`)
    }

    const { config } = Definition.fullSchema({
      pluginDef: SerializationSchema.deserializedRecord,
    }).definition.parse(data)

    const { plugins, ...userConfig } = config

    return new RichTextV2Definition(
      userConfig,
      plugins.map(({ control }) =>
        control ? { control: { definition: deserializeCallback(control?.definition) } } : {},
      ),
    )
  }

  static fullSchema<S extends SchemaTypeAny>({ pluginDef }: { pluginDef: S }) {
    const baseSchema = super.schema
    const plugin = z.object({
      control: z
        .object({
          definition: pluginDef,
        })
        .optional(),
    })

    const config = baseSchema.userConfig.extend({
      defaultValue: z.string(),
      plugins: z.array(plugin),
    })

    return {
      ...baseSchema,
      config,
      definition: z.object({
        type: baseSchema.type,
        config,
      }),
    }
  }

  get configSchema(): SchemaType<Config> {
    return Definition.fullSchema({ pluginDef: z.any() as SchemaType<ControlDefinition> }).config
  }

  createInstance(sendMessage: SendMessage): InstanceType {
    return new RichTextV2Control(sendMessage, this)
  }

  resolveValue(
    data: DataType | undefined,
    _resolver: ResourceResolver,
    _stylesheet: Stylesheet,
    control?: InstanceType,
  ): Resolvable<ReactNode | undefined> {
    const stableValue = StableValue({
      name: Definition.type,
      read: () => renderRichTextV2(data, this, control ?? null),
    })

    return {
      ...stableValue,
      triggerResolve: async () => {},
    }
  }

  getTranslatableData(data: DataType | undefined): Data {
    if (data == null) return null
    return getTranslatableData(Definition.dataToNodes(data), this.config.plugins)
  }

  mergeTranslatedData(
    data: DataType | undefined,
    translatedData: Data,
    _context: MergeTranslatableDataContext,
  ): Data {
    if (data == null || translatedData == null) return data as Data

    const { descendants, ...rest } = Definition.normalizeData(data)
    return {
      ...rest,
      descendants: mergeTranslatedNodes(
        descendants,
        translatedData as RichTextTranslationDto,
        this.config.plugins,
      ),
    }
  }

  serialize(): [SerializedRecord, Transferable[]] {
    const { plugins, ...config } = this.config

    // serialize only the plugin control definition, if any
    const pluginDefs = plugins.map(({ control }) =>
      control
        ? {
            control: {
              definition: control.definition,
              // FIXME: remove getValue/onChange stubs once we released a version of the builder
              // built against the runtime where these can be optional
              getValue: () => undefined,
              onChange: () => {},
            },
          }
        : {},
    )

    return serialize(
      { ...config, plugins: pluginDefs },
      {
        type: Definition.type,
      },
    )
  }

  get pluginControls(): RichTextPluginControl[] {
    return this.config.plugins.map(plugin => plugin.control).filter(isNotNil)
  }

  pluginControlAt(index: number): RichTextPluginControl | undefined {
    return this.pluginControls[index]
  }

  toText(data: DataType | undefined): string {
    if (data == null) return ''
    return toText(Definition.dataToNodes(data), this.config.mode ?? Definition.Mode.Block)
  }
}

export class RichTextV2Definition extends Definition {}

export function RichText(config?: UserConfig): RichTextV2Definition {
  return new RichTextV2Definition(config ?? {})
}

RichText.Mode = Definition.Mode
RichText.isV1Data = Definition.isV1Data
RichText.dataToNodes = Definition.dataToNodes
RichText.Plugin = Plugin

export { RichTextMode as RichTextV2Mode, type DataV2Type as RichTextDataV2, RichTextV2Control }
