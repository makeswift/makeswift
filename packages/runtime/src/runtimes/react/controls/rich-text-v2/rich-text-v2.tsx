import { ReactNode, lazy } from 'react'

import { type DataType } from '@makeswift/controls'
import {
  RichTextV2Control,
  RichTextV2Definition,
  RichTextDataV2,
} from '../../../../controls/rich-text-v2'
import { useIsPreview } from '../../../react/hooks/use-is-preview'
import ReadOnlyTextV2 from './ReadOnlyTextV2'
import ReadOnlyTextV1 from '../rich-text/ReadOnlyText'

export type RichTextV2ControlValue = ReactNode

export type Descriptors = { text?: RichTextV2Definition }

const EditableText = lazy(() => import('./EditableTextV2'))

export function renderRichTextV2(
  data: DataType<RichTextV2Definition> | undefined,
  definition: RichTextV2Definition,
  control: RichTextV2Control | null,
): ReactNode {
  return RichTextV2Definition.isV1Data(data) ? (
    <ReadOnlyTextV1 text={data} />
  ) : (
    <RichTextV2 data={data} definition={definition} control={control} />
  )
}

function RichTextV2({
  data,
  definition,
  control,
}: {
  data: RichTextDataV2 | undefined
  definition: RichTextV2Definition
  control: RichTextV2Control | null
}) {
  return useIsPreview() ? (
    <EditableText text={data} definition={definition} control={control} />
  ) : (
    <ReadOnlyTextV2 text={data} definition={definition} />
  )
}
