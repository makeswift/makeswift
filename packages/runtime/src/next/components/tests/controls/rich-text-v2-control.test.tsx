/** @jest-environment jsdom */
import { RichText } from '../../../../controls'
import { APIResourceType } from '../../../../api'

import * as Fixture from './fixtures/rich-text-v2'
import { testPageControlPropRendering } from './page-control-prop-rendering'

describe('Page', () => {
  describe('RichTextV2', () => {
    test.each([RichText.Mode.Inline, RichText.Mode.Block])(
      `renders a %s placeholder when empty`,
      async mode => {
        await testPageControlPropRendering(RichText({ mode }), {
          value: undefined,
          expectedRenders: 1,
        })
      },
    )

    test(`renders provided text content`, async () => {
      const swatchId = 'U3dhdGNoOmJkODYxMWM5LTNiZjItNDM3MS1iMmU4LTBmMmNlMDZjNDE1OA=='
      const typographyId = 'VHlwb2dyYXBoeTowNGI4OTZlMC0wZWEyLTRkMTMtYmU3ZS0xYmY1M2VmMjBiZjc='

      await testPageControlPropRendering(RichText(), {
        value: Fixture.data,
        cacheData: {
          apiResources: {
            Swatch: [
              {
                id: swatchId,
                value: {
                  __typename: APIResourceType.Swatch,
                  id: swatchId,
                  hue: 238,
                  saturation: 87,
                  lightness: 49,
                },
              },
            ],
            Typography: [
              {
                id: typographyId,
                value: {
                  __typename: APIResourceType.Typography,
                  id: typographyId,
                  name: 'Body',
                  style: [
                    {
                      deviceId: 'desktop',
                      value: {
                        fontFamily: 'Lato',
                        fontSize: { value: 16, unit: 'px' },
                        color: null,
                        lineHeight: null,
                        letterSpacing: null,
                        fontWeight: null,
                        textAlign: null,
                        uppercase: null,
                        underline: null,
                        strikethrough: null,
                        italic: null,
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        expectedRenders: 1,
      })
    })
  })
})
