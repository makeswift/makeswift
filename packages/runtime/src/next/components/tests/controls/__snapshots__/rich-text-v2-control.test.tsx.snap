// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page RichTextV2 renders a makeswift::controls::rich-text-v2::mode::block placeholder when empty: resolvedValue 1`] = `
NodeList [
  .emotion-0 {
  display: inline-block;
  width: 0;
  max-width: 100%;
  white-space: nowrap;
  opacity: 0.333;
  vertical-align: text-top;
}

<div
    style="position: relative; white-space: pre-wrap; word-wrap: break-word;"
  >
    <span
      class="emotion-0"
    >
      Write some text...
    </span>
  </div>,
]
`;

exports[`Page RichTextV2 renders a makeswift::controls::rich-text-v2::mode::block placeholder when empty: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": undefined,
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page RichTextV2 renders a makeswift::controls::rich-text-v2::mode::inline placeholder when empty: resolvedValue 1`] = `
NodeList [
  .emotion-0 {
  display: inline-block;
  width: 0;
  max-width: 100%;
  white-space: nowrap;
  opacity: 0.333;
  vertical-align: text-top;
}

<div
    style="position: relative; white-space: pre-wrap; word-wrap: break-word;"
  >
    <span
      class="emotion-0"
    >
      Write some text...
    </span>
  </div>,
]
`;

exports[`Page RichTextV2 renders a makeswift::controls::rich-text-v2::mode::inline placeholder when empty: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {},
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": undefined,
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;

exports[`Page RichTextV2 renders provided text content: resolvedValue 1`] = `
NodeList [
  .emotion-0 {
  margin: 0;
}

.emotion-2 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

@media only screen and (min-width: 769px) {
  .emotion-3 {
    color: hsla(238,87%,49%,1);
    font-family: Lato;
    font-size: 16px;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-3 {
    color: hsla(238,87%,49%,1);
    font-family: Lato;
    font-size: 16px;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-3 {
    color: hsla(238,87%,49%,1);
    font-family: Lato;
    font-size: 16px;
  }
}

<div
    style="position: relative; white-space: pre-wrap; word-wrap: break-word;"
  >
    <p
      class="emotion-0"
    >
      <span
        class="emotion-1"
      >
        ﻿
      </span>
      <a
        class="emotion-2"
        href="https://www.bigcommerce.com/adyen"
        target="_blank"
      >
        <span
          class="emotion-3"
        >
          Adyen
        </span>
      </a>
      <span
        class="emotion-1"
      >
        ﻿
      </span>
    </p>
    <p
      class="emotion-0"
    >
      <span
        class="emotion-1"
      >
        ﻿
      </span>
      <a
        class="emotion-2"
        href="http://www.authorize.net/"
        target="_blank"
      >
        <span
          class="emotion-3"
        >
          Authorize.net
        </span>
      </a>
      <span
        class="emotion-1"
      >
        ﻿
      </span>
    </p>
    <p
      class="emotion-0"
    >
      <span
        class="emotion-1"
      >
        ﻿
      </span>
      <a
        class="emotion-2"
        href="https://www.bigcommerce.com/bluesnap/"
        target="_blank"
      >
        <span
          class="emotion-3"
        >
          BlueSnap
        </span>
      </a>
      <span
        class="emotion-1"
      >
        ﻿
      </span>
    </p>
    <p
      class="emotion-0"
    >
      <span
        class="emotion-1"
      >
        ﻿
      </span>
      <a
        class="emotion-2"
        href="https://support.bigcommerce.com/s/article/Connecting-with-Checkoutcom/"
        target="_blank"
      >
        <span
          class="emotion-3"
        >
          Checkout.com
        </span>
      </a>
      <span
        class="emotion-1"
      >
        ﻿
      </span>
    </p>
    <p
      class="emotion-0"
    >
      <span
        class="emotion-1"
      >
        ﻿
      </span>
      <a
        class="emotion-2"
        href="http://www.cybersource.com/"
        target="_blank"
      >
        <span
          class="emotion-3"
        >
          CyberSource Direct
        </span>
      </a>
      <span
        class="emotion-1"
      >
        ﻿
      </span>
    </p>
  </div>,
]
`;

exports[`Page RichTextV2 renders provided text content: snapshot 1`] = `
{
  "cacheData": {
    "apiResources": {
      "Swatch": [
        {
          "id": "U3dhdGNoOmJkODYxMWM5LTNiZjItNDM3MS1iMmU4LTBmMmNlMDZjNDE1OA==",
          "value": {
            "__typename": "Swatch",
            "hue": 238,
            "id": "U3dhdGNoOmJkODYxMWM5LTNiZjItNDM3MS1iMmU4LTBmMmNlMDZjNDE1OA==",
            "lightness": 49,
            "saturation": 87,
          },
        },
      ],
      "Typography": [
        {
          "id": "VHlwb2dyYXBoeTowNGI4OTZlMC0wZWEyLTRkMTMtYmU3ZS0xYmY1M2VmMjBiZjc=",
          "value": {
            "__typename": "Typography",
            "id": "VHlwb2dyYXBoeTowNGI4OTZlMC0wZWEyLTRkMTMtYmU3ZS0xYmY1M2VmMjBiZjc=",
            "name": "Body",
            "style": [
              {
                "deviceId": "desktop",
                "value": {
                  "color": null,
                  "fontFamily": "Lato",
                  "fontSize": {
                    "unit": "px",
                    "value": 16,
                  },
                  "fontWeight": null,
                  "italic": null,
                  "letterSpacing": null,
                  "lineHeight": null,
                  "strikethrough": null,
                  "textAlign": null,
                  "underline": null,
                  "uppercase": null,
                },
              },
            ],
          },
        },
      ],
    },
    "localizedResourcesMap": {},
  },
  "document": {
    "data": {
      "key": "00000000-0000-0000-0000-000000000000",
      "props": {
        "children": {
          "columns": [
            {
              "deviceId": "desktop",
              "value": {
                "count": 12,
                "spans": [
                  [
                    12,
                  ],
                ],
              },
            },
          ],
          "elements": [
            {
              "key": "11111111-1111-1111-1111-111111111111",
              "props": {
                "propKey": {
                  "descendants": [
                    {
                      "children": [
                        {
                          "text": "",
                        },
                        {
                          "children": [
                            {
                              "text": "Adyen",
                              "typography": {
                                "id": "VHlwb2dyYXBoeTowNGI4OTZlMC0wZWEyLTRkMTMtYmU3ZS0xYmY1M2VmMjBiZjc=",
                                "style": [
                                  {
                                    "deviceId": "desktop",
                                    "value": {
                                      "color": {
                                        "alpha": 1,
                                        "swatchId": "U3dhdGNoOmJkODYxMWM5LTNiZjItNDM3MS1iMmU4LTBmMmNlMDZjNDE1OA==",
                                      },
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                          "link": {
                            "payload": {
                              "openInNewTab": true,
                              "url": "https://www.bigcommerce.com/adyen",
                            },
                            "type": "OPEN_URL",
                          },
                          "type": "link",
                        },
                        {
                          "text": "",
                        },
                      ],
                      "type": "paragraph",
                    },
                    {
                      "children": [
                        {
                          "text": "",
                        },
                        {
                          "children": [
                            {
                              "text": "Authorize.net",
                              "typography": {
                                "id": "VHlwb2dyYXBoeTowNGI4OTZlMC0wZWEyLTRkMTMtYmU3ZS0xYmY1M2VmMjBiZjc=",
                                "style": [
                                  {
                                    "deviceId": "desktop",
                                    "value": {
                                      "color": {
                                        "alpha": 1,
                                        "swatchId": "U3dhdGNoOmJkODYxMWM5LTNiZjItNDM3MS1iMmU4LTBmMmNlMDZjNDE1OA==",
                                      },
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                          "link": {
                            "payload": {
                              "openInNewTab": true,
                              "url": "http://www.authorize.net/",
                            },
                            "type": "OPEN_URL",
                          },
                          "type": "link",
                        },
                        {
                          "text": "",
                        },
                      ],
                      "type": "paragraph",
                    },
                    {
                      "children": [
                        {
                          "text": "",
                        },
                        {
                          "children": [
                            {
                              "text": "BlueSnap",
                              "typography": {
                                "id": "VHlwb2dyYXBoeTowNGI4OTZlMC0wZWEyLTRkMTMtYmU3ZS0xYmY1M2VmMjBiZjc=",
                                "style": [
                                  {
                                    "deviceId": "desktop",
                                    "value": {
                                      "color": {
                                        "alpha": 1,
                                        "swatchId": "U3dhdGNoOmJkODYxMWM5LTNiZjItNDM3MS1iMmU4LTBmMmNlMDZjNDE1OA==",
                                      },
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                          "link": {
                            "payload": {
                              "openInNewTab": true,
                              "url": "https://www.bigcommerce.com/bluesnap/",
                            },
                            "type": "OPEN_URL",
                          },
                          "type": "link",
                        },
                        {
                          "text": "",
                        },
                      ],
                      "type": "paragraph",
                    },
                    {
                      "children": [
                        {
                          "text": "",
                        },
                        {
                          "children": [
                            {
                              "text": "Checkout.com",
                              "typography": {
                                "id": "VHlwb2dyYXBoeTowNGI4OTZlMC0wZWEyLTRkMTMtYmU3ZS0xYmY1M2VmMjBiZjc=",
                                "style": [
                                  {
                                    "deviceId": "desktop",
                                    "value": {
                                      "color": {
                                        "alpha": 1,
                                        "swatchId": "U3dhdGNoOmJkODYxMWM5LTNiZjItNDM3MS1iMmU4LTBmMmNlMDZjNDE1OA==",
                                      },
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                          "link": {
                            "payload": {
                              "openInNewTab": true,
                              "url": "https://support.bigcommerce.com/s/article/Connecting-with-Checkoutcom/",
                            },
                            "type": "OPEN_URL",
                          },
                          "type": "link",
                        },
                        {
                          "text": "",
                        },
                      ],
                      "type": "paragraph",
                    },
                    {
                      "children": [
                        {
                          "text": "",
                        },
                        {
                          "children": [
                            {
                              "text": "CyberSource Direct",
                              "typography": {
                                "id": "VHlwb2dyYXBoeTowNGI4OTZlMC0wZWEyLTRkMTMtYmU3ZS0xYmY1M2VmMjBiZjc=",
                                "style": [
                                  {
                                    "deviceId": "desktop",
                                    "value": {
                                      "color": {
                                        "alpha": 1,
                                        "swatchId": "U3dhdGNoOmJkODYxMWM5LTNiZjItNDM3MS1iMmU4LTBmMmNlMDZjNDE1OA==",
                                      },
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                          "link": {
                            "payload": {
                              "openInNewTab": true,
                              "url": "http://www.cybersource.com/",
                            },
                            "type": "OPEN_URL",
                          },
                          "type": "link",
                        },
                        {
                          "text": "",
                        },
                      ],
                      "type": "paragraph",
                    },
                  ],
                  "key": "c438a1d3-fee2-4eb4-b49b-f1856c12de42",
                  "type": "makeswift::controls::rich-text-v2",
                  "version": 2,
                },
              },
              "type": "TestComponent",
            },
          ],
        },
      },
      "type": "./components/Root/index.js",
    },
    "fonts": [],
    "id": "test-page-id",
    "locale": null,
    "localizedPages": [],
    "meta": {},
    "seo": {},
    "site": {
      "id": "test-site-id",
    },
    "snippets": [],
  },
}
`;
